### ISSUE TITLE
Add rule to find faulty state derivation in react components

### ISSUE THREAD

**Issue by marutypes** [Shopify/eslint-plugin-shopify#179]
## Problem

Sometimes components derive state from props using  class property initializer syntax

```
class BrokenModal extends React.Component {
  state = {
    open = this.props.open
  }

  render() {
    ...
  }
}
```

Which is fine, however if they don't have any of `static getDerivedStateFromProps`, `componentWillUpdate`, or `componentDidUpdate`, they will go out of synch with their expected behaviour if their props are updated. This can be very difficult and time consuming to debug when encountered trying to use a component the developer didn't write.

The same thing can also happen with state assigned in the constructor.

## Proposal

We should flag components with the following characteristics as invalid:
- state initialized from props in a property initializer OR in the constructor
- no `componentWillUpdate`, `componentDidUpdate`, `static getDerivedStateFromProps`, or `componentWillReceiveProps`

**Comment by alexandcote** [comment]
This issue is 2 years old and we encourage building functional components with hooks now. You can reopen this issue and transfer it to the [web-foundation](https://github.com/shopify/web-foundation) if you think we should fix this.
